### PR TITLE
feat: proof of work audit part 3

### PR DESCRIPTION
**Pull request by hansieodendaal**
Description
---
- Removed `max_block_time` for `struct LinearWeightedMovingAverage`:
  - Removed manual assignment of `max_block_time` for `struct LinearWeightedMovingAverage` as it is always a fixed multiple of `target_time` according to the **LWMA-1** algorithm specification. 
  - Previously, `max_block_time` was asserted when compiling tests and debug mode with function `fn assert_hybrid_pow_constants(`. 
  - This change resulted in the corresponding removal of  `max_target_time` from `struct PowAlgorithmConstants` (and consensus constants) as it is now being calculated by `LinearWeightedMovingAverage`.
- Added documentation to public methods.

Dependent on #5495 

Motivation and Context
---
Preparation for code audit

How Has This Been Tested?
---
All unit tests pass

What process can a PR reviewer use to test or verify this change?
---
Code walkthrough

Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

### CODE REVIEW

**Comment by stringhandler** [comment]
Needs more description about the difficulty changes

File: base_layer/core/src/consensus/consensus_constants.rs [hunk 1]
```diff
@@ -243,11 +242,9 @@ impl ConsensusConstants {
 
     /// The maximum time a block is considered to take. Used by the difficulty adjustment algorithms
     /// Multiplied by the PoW algorithm block percentage.
-    pub fn pow_max_block_interval(&self, pow_algo: PowAlgorithm) -> u64 {
-        match self.proof_of_work.get(&pow_algo) {
-            Some(v) => v.max_target_time,
-            _ => 0,
-        }
+    pub fn pow_max_block_interval(&self, pow_algo: PowAlgorithm) -> Result<u64, String> {
+        LinearWeightedMovingAverage::max_block_time(self.pow_target_block_interval(pow_algo))
```
Comment (stringhandler) [inline base_layer/core/src/consensus/consensus_constants.rs]: This seems like a big change, possibly leaking out the implementation that's abstracted by proof of work. Can you give some more reasons as to this change?

File: base_layer/core/src/proof_of_work/error.rs [hunk 2]
```diff
@@ -48,7 +48,9 @@ pub enum DifficultyAdjustmentError {
 }
 
 #[derive(Debug, Error)]
-pub enum DifficultyError {
+pub enum LwmaError {
```
Comment (stringhandler) [inline base_layer/core/src/proof_of_work/error.rs]: This does not seem like a good renaming. The variant "InvalidDifficulty" does not seem related to Lwma at all

File: base_layer/core/src/validation/error.rs [hunk 3]
```diff
@@ -152,7 +152,7 @@ pub enum ValidationError {
     #[error("Not enough timestamps provided. Expected {expected}, got {actual}")]
     NotEnoughTimestamps { expected: usize, actual: usize },
     #[error("Invalid difficulty: {0}")]
-    DifficultyError(#[from] DifficultyError),
+    LwmaError(#[from] LwmaError),
```
Comment (stringhandler) [inline base_layer/core/src/validation/error.rs]: I also don't really like this rename

Comment (stringhandler) [reply base_layer/core/src/validation/error.rs]: Comment is wrong if nothing else

**Review by stringhandler — CHANGES_REQUESTED** [review]

File: base_layer/core/src/consensus/consensus_constants.rs (hunk 1 above)
Comment (hansieodendaal) [reply base_layer/core/src/consensus/consensus_constants.rs]: I propose updating the comment as:
```
    /// The maximum time/interval a block is considered to take. This is used by the difficulty adjustment algorithms, 
    /// and multiplied by the PoW algorithm block percentage. For LWMA-1, `max_block_time` is a fixed multiple of 
    /// `target_block_time`, determined by `LinearWeightedMovingAverage`.
```

Comment (stringhandler) [reply base_layer/core/src/consensus/consensus_constants.rs]: it's more about duplicating code in a strange place that is unlikely to be found. I preferred the approach of asking `self.proof_of_work` for this answer

Comment (hansieodendaal) [reply base_layer/core/src/consensus/consensus_constants.rs]: After discussion with @stringhandler, accepting the suggestion to remove this function as it is easily calculated with `LinearWeightedMovingAverage::max_block_time(target_time)` and passing in `ConsensusConstants::pow_target_block_interval(..)`, as `max_block_time` is not a consensus parameter.

File: applications/tari_app_grpc/proto/types.proto [hunk 4]
```diff
@@ -64,6 +64,7 @@ message CommitmentSignature {
 
 /// PoW Algorithm constants
 message PowAlgorithmConstants {
+    uint64 max_target_time = 1;
```
Comment (stringhandler) [inline applications/tari_app_grpc/proto/types.proto]: This does not exist in the struct anymore, so can be removed

File: base_layer/core/src/mempool/shrink_hashmap.rs [hunk 5]
```diff
@@ -20,8 +20,10 @@
 //  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 //  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+#[cfg(feature = "base_node")]
```
Comment (SWvheerden) [inline base_layer/core/src/mempool/shrink_hashmap.rs]: put this in the mod file rather

Comment (SWvheerden) [reply base_layer/core/src/mempool/shrink_hashmap.rs]: the cfg flag, for entire file

**Review by SWvheerden — APPROVED** [review]
